model: Added 3 HIT-TMG's KaLM-embedding models

ayush1298 opened the pull request:
fixes #1445 #2482 
Added 3 models: 
1. HIT-TMG/KaLM-embedding-multilingual-mini-instruct-v1 with instruct wrapper 
2. HIT_TMG/KaLM_embedding_multilingual_mini_v1 
3. HIT-TMG/KaLM-embedding-multilingual-mini-instruct-v1.5 with instruct wrapper 

### Code Quality
<!-- Please do not delete this -->
- [x] **Code Formatted**: Format the code using `make lint` to maintain consistent style.

### Documentation
<!-- Please do not delete this -->
- [ ] **Updated Documentation**: Add or update documentation to reflect the changes introduced in this PR.

### Testing
<!-- Please do not delete this -->
- [ ] **New Tests Added**: Write tests to cover new functionality. Validate with `make test-with-coverage`.
- [x] **Tests Passed**: Run tests locally using `make test` or `make test-with-coverage` to ensure no existing functionality is broken.

### Adding a model checklist
<!--
When adding a model to the model registry
see also https://github.com/embeddings-benchmark/mteb/blob/main/docs/reproducible_workflow.md
-->

 - [x] I have filled out the ModelMeta object to the extent possible
 - [x] I have ensured that my model can be loaded using
   - [x] `mteb.get_model(model_name, revision)` and
   - [x] `mteb.get_model_meta(model_name, revision)`
 - [ ] I have tested the implementation works on a representative set of tasks.

ayush1298 commented:
@Samoed I will not be able to run the models on all tasks and add results to results repo. Can you do that if possible?

Samoed commented on mteb/models/hit_tmg_models.py:
```diff
@@ -0,0 +1,168 @@
+HIT_TMG__KaLM_embedding_multilingual_mini_v1 = ModelMeta(
+    loader=partial(  # type: ignore
+        InstructSentenceTransformerWrapper,
+        model_name_or_path="HIT-TMG/KaLM-embedding-multilingual-mini-v1",
```
I don't think that this soulbe be instructed

ayush1298 replied:
I thought in this [comment](https://github.com/embeddings-benchmark/mteb/issues/1445#issuecomment-2772962434), you were saying to add both.

Samoed replied:
Hm, I've missed that this is not instruct model

ayush1298 replied:
So, I have to remove this loader only, right?

Samoed replied:
Yes, only loader and instruction template

ayush1298 replied:
I have added the commit after this. Will try to add results for some tasks. But, not sure how much I will be able to as I don't have GPU. Can this be done in google colab?

Samoed replied:
I think yes. I have kaggle/colab resources only too

ayush1298 replied:
@Samoed When I am loading the models using command:
```
import mteb

model_name = "HIT-TMG/KaLM-embedding-multilingual-mini-instruct-v1"
revision = "45e42c89990c40aca042659133fc8b13c28634b5"
model = mteb.get_model(model_name=model_name, revision=revision, trust_remote_code= True)
```
then it is giving warning as: 
```
WARNING:mteb.model_meta:Loader not specified for model HIT-TMG/KaLM-embedding-multilingual-mini-instruct-v1, loading using sentence transformers.
```
Why I am getting this though loader is specified for `HIT-TMG/KaLM-embedding-multilingual-mini-instruct-v1`
 

Samoed replied:
> When I am loading the models using command

You need to add model to `models/overview.py`

> Also, on which tasks should I run these 3 models?

They have added results to their model card  https://github.com/embeddings-benchmark/results/tree/main/results/HIT-TMG__KaLM-embedding-multilingual-mini-instruct-v1/external you can compare your implementation with their results

ayush1298 replied:
I don't have to push results right?
Will it be okay, if I just check results on some tasks?
For STS12, following is comparison table: 
# **Comparison of New and Old Implementations**

| Metric               | New Implementation | Old Implementation |
|----------------------|-------------------|-------------------|
| **Pearson**         | 0.87527           | 0.87941           |
| **Spearman**        | 0.810452          | 0.81495           |
| **Cosine Pearson**  | 0.87527           | 0.87941           |
| **Cosine Spearman** | 0.810453          | 0.81495           |
| **Manhattan Pearson** | 0.858329        | 0.85794           |
| **Manhattan Spearman** | 0.812035       | 0.81688           |
| **Euclidean Pearson** | 0.857291        | 0.85671           |
| **Euclidean Spearman** | 0.810451       | 0.81495           |
| **Main Score**      | 0.810453          | 0.81495           |

Samoed replied:
This is the result of which model? Can you check results for one task for each task type of submitted results for every model?

ayush1298 replied:
Just updated the above comment, its result for `HIT-TMG/KaLM-embedding-multilingual-mini-instruct-v1` model

ayush1298 replied:
> for each task type
>
What do you mean by this exactly?

Samoed replied:
Select 1 task from submitted results to run for STS, classification, rertrieval etc

ayush1298 replied:
Okay, will do it. Should I add all these comparison results in PR description. And are the above results are correct? There is some deviation but I think that was in 2nd or 3rd decimal places mostly.

Samoed replied:
> Should I add all these comparison results in PR description

Yes, or to separete comment.

> There is some deviation but I think that was in 2nd or 3rd decimal places mostly.

This is normal, especially when comapring with external results

ayush1298 replied:
Do you have any list of tasks for each task types to run, which are easy to evaluate and did not get Cuda Out of Memory error. I have tried 4-5 different tasks but getting this error.

Samoed replied:
I don't think Cuda oom related to tasks. Can you try to run on kaggle and share notebook if you have same problems?

Samoed commented on mteb/models/hit_tmg_models.py:
```diff
@@ -0,0 +1,168 @@
+    similarity_fn_name="cosine",
+    use_instructions=None,
+    training_datasets=kalm_training_data,
+    adapted_from="/mnt/shgeminicephfs/wx-dc-plt-hpc/xinshuohu/Output/Embedding/Qwen2-0.5B-eos_mean_pretrain_0806_1e-4_uen_sft_0902_filtered_v2_3node_g8_1e-5_sin-0.1",
```
```suggestion
    adapted_from="Qwen/Qwen2-0.5B",
```

Samoed commented on mteb/models/hit_tmg_models.py:
```diff
@@ -0,0 +1,168 @@
+    similarity_fn_name="cosine",
+    use_instructions=True,
+    training_datasets=kalm_training_data,  # Replace with actual dataset if available
+    adapted_from="/mnt/shgeminicephfs/wx-dc-plt-hpc/xinshuohu/Output/Embedding/Qwen2-0.5B-eos_mean_pretrain_0806_1e-4_uen_sft_1022_filtered_v2_inst_3node_g8_1e-5_sin-0.1_mrl",
```
```suggestion
    adapted_from="Qwen/Qwen2-0.5B",
```

Samoed commented on mteb/models/hit_tmg_models.py:
```diff
@@ -0,0 +1,168 @@
+    open_weights=True,
+    public_training_code=None,
+    public_training_data=None,
+    framework=["PyTorch"],
```
```suggestion
    framework=["PyTorch", "Sentence Transformers"],
```

Samoed reviewed (commented):
Can you also add https://huggingface.co/HIT-TMG/KaLM-embedding-multilingual-mini-instruct-v1.5?

ayush1298 commented on mteb/models/ops_moa_models.py:
```diff
@@ -2,6 +2,7 @@
 
 from functools import partial
 
+import numpy as np
```
Added because `make lint` command was giving error.

ayush1298 commented on docs/tasks.md:
```diff
@@ -905,7 +905,7 @@ The following tables give you an overview of the tasks in MTEB.
 <!-- TASK LANG TABLE START -->
 | ISO Code | Language | Family | Any2AnyMultiChoice | Any2AnyMultilingualRetrieval | Any2AnyRetrieval | BitextMining | Classification | Clustering | Compositionality | DocumentUnderstanding | ImageClassification | ImageClustering | ImageMultilabelClassification | InstructionRetrieval | MultilabelClassification | PairClassification | Reranking | Retrieval | STS | Speed | Summarization | VisionCent ...
-|---|------|------|------|------|------|------|------|------|------|------|------|------|------|------|------|------|------|------|------|------|------|------|------|---|
+|----------|---------|--------|--------------------|-----------------------------|------------------|-------------|---------------|------------|-----------------|----------------------|---------------------|----------------|----------------------------|-------------------|---------------------|-----------------|----------|-----------|-----|-------|--------------|---------------|----------------|- ...
```
@Samoed Added this as I found out the table was not showing correctly. But after this change, that is fixed. But don't know why it is showing merged conflicts with tasks.md

ayush1298 replied:
I see that its just changing of name from `VisionCentric` to `VisionCentricQA`. Should I also rename that, so it will not cause merge conflict, or add this in separate PR after fetching latest main?

Samoed replied:
You can just rename it

ayush1298 commented:
Detailed Analysis of Results Comparison: 
M1: HIT-TMG/KaLM-embedding-multilingual-mini-instruct-v1
M2: HIT-TMG/KaLM-embedding-multilingual-mini-instruct-v1.5
M3: HIT-TMG/KaLM-embedding-multilingual-mini-v1

<details>
<summary> Task Type - "Classification": Task - "EmotionClassification" </summary>
Significant Differences in Results

| Metric        | M1-New  | M1-Old  | M2-New  | M2-Old  | M3-New  | M3-Old  |
|--------------|--------|--------|--------|--------|--------|--------|
| **Accuracy** | 0.604  | 0.85565 | 0.6017 | 0.869  | 0.5118 | 0.53945 |
| **F1**       | 0.5475 | 0.81123 | 0.5469 | 0.82434 | 0.4573 | 0.46749 |
| **F1 Weighted** | 0.6202 | 0.85983 | 0.6184 | 0.87211 | 0.5321 | 0.55445 |
| **Main Score** | 0.604  | 0.85565 | 0.6017 | 0.869  | 0.5118 | 0.53945 |
</details> 

<details>
<summary> Task Type - "MultilabelClassification": Task - "CEDRClassification" </summary>

| Metric        | M1-New  | M1-Old  | M2-New  | M2-Old  | M3-New  | M3-Old  |
|--------------|--------|--------|--------|--------|--------|--------|
| **Accuracy** | 0.3972 | 0.4330 | 0.3908 | 0.4376 | 0.4015 | 0.4216 |
| **F1**       | 0.2724 | 0.4111 | 0.2719 | 0.4247 | 0.3118 | 0.3909 |
| **LRAP**     | 0.6429 | 0.7206 | 0.6409 | 0.7363 | 0.6595 | 0.7107 |
| **Main Score** | 0.3972 | 0.4330 | 0.3908 | 0.4376 | 0.4015 | 0.4216 |
</details> 


<details>
<summary> Task Type - "Clustering": Task - "GeoreviewClusteringP2P" </summary>

| Metric          | M1-New  | M1-Old  | M2-New  | M2-Old  | M3-New  | M3-Old  |
|----------------|--------|--------|--------|--------|--------|--------|
| **Main Score** | 0.6329 | 0.6028 | 0.6324 | 0.6076 | 0.6211 | 0.6340 |
| **V-Measure**  | 0.6329 | 0.6028 | 0.6324 | 0.6076 | 0.6211 | 0.6340 |
| **V-Measure Std** | 0.0088 | 0.0103 | 0.0091 | 0.0066 | 0.0098 | 0.0044 |
</details> 

<details>
<summary> Task Type - "PairClassification": Task - "Ocnli" </summary>
<html><head></head><body>
Metric | Model 1 Old | Model 1 New | Model 2 Old | Model 2 New | Model 3 Old | Model 3 New
-- | -- | -- | -- | -- | -- | --
Cosine Accuracy | 0.6687 | 0.6665 | 0.6622 | 0.6622 | 0.6703 | 0.6703
Cosine Accuracy Threshold | 0.8435 | 0.8587 | 0.8553 | 0.8553 | 0.6514 | 0.6514
Cosine AP | 0.6983 | 0.6968 | 0.6921 | 0.6921 | 0.6949 | 0.6949
Cosine F1 | 0.7134 | 0.7075 | 0.7045 | 0.7045 | 0.7128 | 0.7128
Cosine F1 Threshold | 0.8386 | 0.8333 | 0.8385 | 0.8385 | 0.5884 | 0.5884
Cosine Precision | 0.6393 | 0.6043 | 0.6136 | 0.6136 | 0.6234 | 0.6234
Cosine Recall | 0.8068 | 0.8532 | 0.8268 | 0.8268 | 0.8321 | 0.8321
Dot Accuracy | 0.6687 | 0.6665 | 0.6622 | 0.6622 | 0.6703 | 0.6703
Dot Accuracy Threshold | 0.8435 | 0.8587 | 0.8553 | 0.8553 | 0.6514 | 0.6514
Dot AP | 0.6983 | 0.6968 | 0.6921 | 0.6921 | 0.6949 | 0.6949
Dot F1 | 0.7134 | 0.7075 | 0.7045 | 0.7045 | 0.7128 | 0.7128
Dot F1 Threshold | 0.8386 | 0.8333 | 0.8385 | 0.8385 | 0.5884 | 0.5884
Dot Precision | 0.6393 | 0.6043 | 0.6136 | 0.6136 | 0.6234 | 0.6234
Dot Recall | 0.8068 | 0.8532 | 0.8268 | 0.8268 | 0.8321 | 0.8321
Euclidean Accuracy | 0.6687 | 0.6665 | 0.6622 | 0.6622 | 0.6703 | 0.6703
Euclidean Accuracy Threshold | 0.5595 | 0.5316 | 0.5379 | 0.5379 | 0.8349 | 0.8349
Euclidean AP | 0.6983 | 0.6968 | 0.6921 | 0.6921 | 0.6949 | 0.6949
Euclidean F1 | 0.7134 | 0.7075 | 0.7045 | 0.7045 | 0.7128 | 0.7128
Euclidean F1 Threshold | 0.5682 | 0.5775 | 0.5682 | 0.5682 | 0.9074 | 0.9074
Euclidean Precision | 0.6393 | 0.6043 | 0.6136 | 0.6136 | 0.6234 | 0.6234
Euclidean Recall | 0.8068 | 0.8532 | 0.8268 | 0.8268 | 0.8321 | 0.8321
Manhattan Accuracy | 0.6605 | 0.6611 | 0.6589 | 0.6589 | 0.6692 | 0.6692
Manhattan Accuracy Threshold | 12.1546 | 12.0459 | 12.5711 | 12.5711 | 19.5184 | 19.5184
Manhattan AP | 0.6951 | 0.6940 | 0.6893 | 0.6893 | 0.6902 | 0.6902
Manhattan F1 | 0.7056 | 0.7068 | 0.7002 | 0.7002 | 0.7090 | 0.7090
Manhattan F1 Threshold | 13.4479 | 13.3175 | 13.2169 | 13.2169 | 22.0142 | 22.0142
Manhattan Precision | 0.6178 | 0.6184 | 0.6172 | 0.6172 | 0.5891 | 0.5891
Manhattan Recall | 0.8226 | 0.8247 | 0.8089 | 0.8089 | 0.8902 | 0.8902
Max AP | 0.6983 | 0.6968 | 0.6921 | 0.6921 | 0.6949 | 0.6949
Max F1 | 0.7134 | 0.7075 | 0.7045 | 0.7045 | 0.7128 | 0.7128
Max Precision | 0.6393 | 0.6184 | 0.6172 | 0.6172 | 0.6234 | 0.6234
Max Recall | 0.8226 | 0.8532 | 0.8268 | 0.8268 | 0.8902 | 0.8902
Similarity Accuracy | 0.6687 | 0.6665 | 0.6622 | 0.6622 | 0.6703 | 0.6703
Similarity Accuracy Threshold | 0.8435 | 0.8587 | 0.8553 | 0.8553 | 0.6514 | 0.6514
Similarity AP | 0.6983 | 0.6968 | 0.6921 | 0.6921 | 0.6949 | 0.6949
Similarity F1 | 0.7134 | 0.7075 | 0.7045 | 0.7045 | 0.7128 | 0.7128
Similarity F1 Threshold | 0.8386 | 0.8333 | 0.8385 | 0.8385 | 0.5884 | 0.5884
Similarity Precision | 0.6393 | 0.6043 | 0.6136 | 0.6136 | 0.6234 | 0.6234
Similarity Recall | 0.8068 | 0.8532 | 0.8268 | 0.8268 | 0.8321 | 0.8321
Main Score | 0.6983 | 0.6665 | 0.6921 | 0.6622 | 0.6949 | 0.6703
</details>

Samoed commented:
Hm. I they've reported different prompts in paper with what've using. Can you update your implementation with their prompts?

ayush1298 commented:
> Hm. I they've reported different prompts in paper with what've using. Can you update your implementation with their prompts? You can change model to use sentence transformer wrapper, but this is a hack and not clear how to integrate their resuls properly. At least can you try to change prompt for 2-3 tasks directly to test if our implementation will match?
>
I think only Classification and MultilabelClassification results are having some differences. For retrieval, reranking, STS tasks, there are no differences.

Samoed commented:
>  I saw thier paper, they have given different task instructions, each 1 is specific for the task. Should we support task-specific instructions in MTEB?

I think you can create an issue for it to discuss. After we will decide what to do with this model

Samoed commented:
@ayush1298 You can change `get_instruction`
https://github.com/embeddings-benchmark/mteb/blob/cb2825ce21a6308fe51fbda5384f6d134b1e3cb1/mteb/models/wrapper.py#L91
similarly to `get_prompt_name` https://github.com/embeddings-benchmark/mteb/blob/cb2825ce21a6308fe51fbda5384f6d134b1e3cb1/mteb/models/wrapper.py#L21

ayush1298 commented:
@Samoed I have modified `get_instruction` similar to `get_prompt_name`, but I don't know how to exactly incorporate this in model_meta for each model.
 
1 more thing, I think what I missed is the prompt given at end in paper are having same format only of:
HIT_TMG_INSTRUCTION = "Instruct: {instruction}\nQuery: "

they just have given these as an example with task-specific instruction and query for each task.

Samoed commented on mteb/models/wrapper.py:
```diff
@@ -92,9 +92,28 @@ def get_instruction(task_name: str, prompt_type: PromptType | None) -> str:
         if isinstance(task_metadata.prompt, dict) and prompt_type:
-            if task_metadata.prompt.get(prompt_type.value):
-                return task_metadata.prompt[prompt_type.value]
+            if (
+                task_name
+                and prompt_type
+                and f"{task_name}-{prompt_type_value}" in prompt_data
+            ):
+                return prompt_data[f"{task_name}-{prompt_type_value}"]
+            if task_name and task_name in prompt_data:
+                return prompt_data[task_name]
+            if (
+                task_type
+                and prompt_type
+                and f"{task_type}-{prompt_type_value}" in prompt_data
+            ):
+                return prompt_data[f"{task_type}-{prompt_type_value}"]
+            if task_type and task_type in prompt_data:
+                return prompt_data[task_type]
+            if prompt_type and prompt_type_value in prompt_data:
+                return prompt_data[prompt_type_value]
```
This won't work for model. You should pass dict of prompts to this function and check is it contains prompts for current task

ayush1298 replied:
>1 more thing, I think what I missed is the prompt given at end in paper are having same format only of:
HIT_TMG_INSTRUCTION = "Instruct: {instruction}\nQuery: "

they just have given these as an example with task-specific instruction and query for each task.
>
As I mentioned, do we really need dict of prompts?

Samoed replied:
> As I mentioned, do we really need dict of prompts?

To replicate their results, yes we need dict of prompts

ayush1298 replied:
@Samoed If we change get_instruction to accept dict of prompts, then what about other models which uses this current implementation without prompt dict like [gme_v_models](https://github.com/embeddings-benchmark/mteb/blob/main/mteb/models/gme_v_models.py) and [llm2vec_models](https://github.com/embeddings-benchmark/mteb/blob/main/mteb/models/llm2vec_models.py) ? Do you want that these function support both?

Samoed replied:
If prompt dict is not passed then default prompt should be used

ayush1298 replied:
@Samoed Can you answer these?

Samoed commented on mteb/models/wrapper.py:
```diff
@@ -88,36 +88,25 @@ def validate_task_to_prompt_name(
+        prompts_dict: dict[str, str] | None = None,
+    ) -> str:
         """Get the instruction/prompt to be used for encoding sentences."""
+        if prompts_dict and task_name in prompts_dict:
```
And what if task want to create different instructions to query and passages?

ayush1298 replied:
What should be done for that?

Samoed replied:
I think code can be changed like
```python
task = mteb.get_task(task_name=task_name)
prompts = task_metadata.prompts
if prompt dict and task_name in prompts_dict:
    prompts = prompts_dict[task_name]
...
```

Samoed commented on mteb/models/hit_tmg_models.py:
```diff
@@ -107,6 +107,68 @@
+    "EightTagsClustering": "Instruct: Identify of headlines from social media posts in Polish into 8 categories: film, history, food, medicine, motorization, work, sport and technology \n Query: {query}",
+    "GeoreviewClusteringP2P": "Instruct: Identify the topic or theme of the Russian reviews. \n Query: {query}",
+    "RuSciBenchGRNTIClusteringP2P": "Instruct: Identify the topic or theme of the Russian articles. \n Query: {query}",
+    "RuSciBenchOECDClusteringP2P": "Instruct: Identify the topic or theme of the Russian articles. \n Query: {query}",
```
You shouldn't add `{query}`, because we append text to instruction

ayush1298 commented:
@Samoed Can you check this? I have added suggested changes
Sorry for late reply.

ayush1298 commented on mteb/models/hit_tmg_models.py:
```diff
@@ -0,0 +1,254 @@
+    "MTOPIntentClassification": ["train"],
+}
+
+HIT_TMG_task_prompts = {
```
Where to exactly pass it, inside `InstructWrapper `call?

Samoed replied:
You can add them to loader as `__init__` params of `InstructSentenceTransformerWrapper`

ayush1298 replied:
so should I add this line under loader,
`prompts_dict = HIT_TMG_task_prompts`

Samoed replied:
Yes and also update `InstructSentenceTransformerWrapper`

Samoed replied:
Great table! Can you also add their reported scores? You can load it's results from results repo https://github.com/embeddings-benchmark/results/tree/main/results/HIT-TMG__KaLM-embedding-multilingual-mini-instruct-v1/external

Samoed replied:
Great table! `Old` is the results reported by `KaLM`? I think currently you don't use prompts from `KaLM`

ayush1298 replied:
Yes old are results reported by KALM. 
```
loader=partial(  # type: ignore
        InstructSentenceTransformerWrapper,
        model_name="HIT-TMG/KaLM-embedding-multilingual-mini-instruct-v1.5",
        revision="fcff2f8a54e4cd96b7766fef1ee960a43d42bb3c",
        instruction_template=HIT_TMG_INSTRUCTION,
        max_seq_length=512,
        apply_instruction_to_passages=True,
        prompts_dict = HIT_TMG_task_prompts,
    ),
```
after these we are using there prompts, right?

Samoed replied:
No, you haven't updated `InstructSentenceTransformerWrapper` to use `prompts_dict`. You can check which instruction is used. It will be reported by https://github.com/embeddings-benchmark/mteb/blob/9cfa2e828bb1ffcc7f96aca848685a92c72bfa47/mteb/models/instruct_wrapper.py#L152

ayush1298 replied:
Can you check it now? If it is correct or not

Samoed replied:
It's a bit wrong because template is applied twice

```
INFO:mteb.models.instruct_wrapper:Using instruction: 'Instruct: Instruct: Classifying the emotion expressed in the given Twitter message into one of the six emotions: anger, fear, joy, love, sadness, and surprise
 Query:
Query: ' for task: 'EmotionClassification'
```

ayush1298 replied:
cool, will remove it. Btw, how did you exactly check these?

Samoed replied:
```bash
mteb run -m HIT-TMG/KaLM-embedding-multilingual-mini-instruct-v1.5 -t EmotionClassification
```

ayush1298 replied:
> You should remove all `Instruct` and `Query` from prompts
>
Done, these one. Will share results on `EmotionClassification`,  `CEDRClassification` , `GeoreviewClusteringP2P` , `Ocnli`?

Samoed replied:
I think it's instruction are applied correctly now
```
INFO:mteb.evaluation.MTEB:Scores: {'default': {'accuracy': 0.86865, 'f1': 0.8239077177532754, 'f1_weighted': 0.8719397109482919, 'scores_per_experiment': [{'accuracy': 0.8705, 'f1': 0.8234721295327191, 'f1_weighted': 0.8727201525356314}, {'accuracy': 0.8705, 'f1': 0.8215391919604418, 'f1_weighted': 0.8719549099435739}, {'accuracy': 0.865, 'f1': 0.8186945389066319, 'f1_weighted': 0.869460776530738}, {'accuracy': 0.8735, 'f1': 0.8244695126532503, 'f1_weighted': 0.8774026951275053}, {'accuracy': 0.8795, 'f1': 0.8381916009889584, 'f1_weighted': 0.8832349691693732}, {'accuracy': 0.865, 'f1': 0.8236566119733676, 'f1_weighted': 0.8679594303317587}, {'accuracy': 0.868, 'f1': 0.8231131688042778, 'f1_weighted': 0.8720979934574868}, {'accuracy': 0.8515, 'f1': 0.8093363809244258, 'f1_weighted': 0.8559114291089968}, {'accuracy': 0.873, 'f1': 0.8333940898313624, 'f1_weighted': 0.8747766206135759}, {'accuracy': 0.87, 'f1': 0.8232099519573198, 'f1_weighted': 0.8738781326642796}], 'main_score': 0.86865}}
```

Samoed replied:
Can you run tasks from https://github.com/embeddings-benchmark/mteb/pull/2478#issuecomment-2776666769 for one instruction model?

ayush1298 commented:
@Samoed these are new tables for all models given in [#2478 (comment)](https://github.com/embeddings-benchmark/mteb/pull/2478#issuecomment-2776666769)
Apart from Ocnil, for rest results are almost matching or very much close to one reported by Kalm.
Detailed Analysis of Results Comparison: 
M1: HIT-TMG/KaLM-embedding-multilingual-mini-instruct-v1
M2: HIT-TMG/KaLM-embedding-multilingual-mini-instruct-v1.5
M3: HIT-TMG/KaLM-embedding-multilingual-mini-v1

<details>
<summary> Task Type - "Classification": Task - "EmotionClassification" </summary>

| Metric        | M1-New  | M1-Old  | M2-New  | M2-Old  | M3-New  | M3-Old  |
|--------------|--------|--------|--------|--------|--------|--------|
| **Accuracy** | 0.86575  | 0.85565 | 0.86865 | 0.869  | 0.512 | 0.53945 |
| **F1**       | 0.821701 | 0.81123 | 0.823851 | 0.82434 | 0.45745 | 0.46749 |
| **F1 Weighted** | 0.869232 | 0.85983 | 0.871923 | 0.87211 | 0.532315 | 0.55445 |
| **Main Score** | 0.86575  | 0.85565 | 0.86865 | 0.869  | 0.512 | 0.53945 |
</details> 

<details>
<summary> Task Type - "MultilabelClassification": Task - "CEDRClassification" </summary>

| Metric        | M1-New  | M1-Old  | M2-New  | M2-Old  | M3-New  | M3-Old  |
|--------------|--------|--------|--------|--------|--------|--------|
| **Accuracy** | 0.438576 | 0.4330 | 0.439107 | 0.4376 | 0.401488| 0.4216 |
| **F1**       | 0.426501 | 0.4111 | 0.427433 | 0.4247 | 0.311745 | 0.3909 |
| **LRAP**     | 0.732202 | 0.7206 | 0.737473 | 0.7363 | 0.659466 | 0.7107 |
| **Main Score** | 0.438576 | 0.4330 | 0.439107 | 0.4376 | 0.401488 | 0.4216 |
</details> 


<details>
<summary> Task Type - "Clustering": Task - "GeoreviewClusteringP2P" </summary>

| Metric          | M1-New  | M1-Old  | M2-New  | M2-Old  | M3-New  | M3-Old  |
|----------------|--------|--------|--------|--------|--------|--------|
| **Main Score** | 0.61396 | 0.6028 | 0.610107 | 0.6076 | 0.629293 | 0.6340 |
| **V-Measure**  | 0.61396 | 0.6028 | 0.610107 | 0.6076 | 0.629293 | 0.6340 |
| **V-Measure Std** | 0.00843 | 0.0103 | 0.007257 | 0.0066 | 0.010543 | 0.0044 |
</details> 

<details>
<summary> Task Type - "PairClassification": Task - "Ocnli" </summary>

| Metric                        | M1-Old | M1-New  | M2-Old | M2-New  | M3-Old | M3-New  |
|------------------------------|--------|---------|--------|---------|--------|---------|
| Cosine Accuracy              | 0.6687 | 0.6302  | 0.6622 | 0.6259  | 0.6703 | 0.6703  |
| Cosine Accuracy Threshold    | 0.8435 | 0.7969  | 0.8553 | 0.7997  | 0.6514 | 0.6514  |
| Cosine AP                    | 0.6983 | 0.6481  | 0.6921 | 0.6471  | 0.6949 | 0.6949  |
| Cosine F1                    | 0.7134 | 0.6868  | 0.7045 | 0.6869  | 0.7128 | 0.7128  |
| Cosine F1 Threshold          | 0.8386 | 0.6921  | 0.8385 | 0.6684  | 0.5884 | 0.5884  |
| Cosine Precision             | 0.6393 | 0.5347  | 0.6136 | 0.5251  | 0.6234 | 0.6234  |
| Cosine Recall                | 0.8068 | 0.9599  | 0.8268 | 0.9926  | 0.8321 | 0.8321  |
| Dot Accuracy                 | 0.6687 | 0.6302  | 0.6622 | 0.6259  | 0.6703 | 0.6703  |
| Dot Accuracy Threshold       | 0.8435 | 0.7969  | 0.8553 | 0.7997  | 0.6514 | 0.6514  |
| Dot AP                       | 0.6983 | 0.6481  | 0.6921 | 0.6471  | 0.6949 | 0.6949  |
| Dot F1                       | 0.7134 | 0.6868  | 0.7045 | 0.6869  | 0.7128 | 0.7128  |
| Dot F1 Threshold             | 0.8386 | 0.6921  | 0.8385 | 0.6684  | 0.5884 | 0.5884  |
| Dot Precision                | 0.6393 | 0.5347  | 0.6136 | 0.5251  | 0.6234 | 0.6234  |
| Dot Recall                   | 0.8068 | 0.9599  | 0.8268 | 0.9926  | 0.8321 | 0.8321  |
| Euclidean Accuracy           | 0.6687 | 0.6302  | 0.6622 | 0.6259  | 0.6703 | 0.6703  |
| Euclidean Accuracy Threshold | 0.5595 | 0.6374  | 0.5379 | 0.6330  | 0.8349 | 0.8349  |
| Euclidean AP                 | 0.6983 | 0.6481  | 0.6921 | 0.6471  | 0.6949 | 0.6949  |
| Euclidean F1                 | 0.7134 | 0.6868  | 0.7045 | 0.6869  | 0.7128 | 0.7128  |
| Euclidean F1 Threshold       | 0.5682 | 0.7848  | 0.5682 | 0.8144  | 0.9074 | 0.9074  |
| Euclidean Precision          | 0.6393 | 0.5347  | 0.6136 | 0.5251  | 0.6234 | 0.6234  |
| Euclidean Recall             | 0.8068 | 0.9599  | 0.8268 | 0.9926  | 0.8321 | 0.8321  |
| Manhattan Accuracy           | 0.6605 | 0.6259  | 0.6589 | 0.6215  | 0.6692 | 0.6692  |
| Manhattan Accuracy Threshold | 12.1546| 14.8894 | 12.5711| 14.8752 | 19.5184| 19.5184 |
| Manhattan AP                 | 0.6951 | 0.6458  | 0.6893 | 0.6445  | 0.6902 | 0.6902  |
| Manhattan F1                 | 0.7056 | 0.6889  | 0.7002 | 0.6872  | 0.7090 | 0.7090  |
| Manhattan F1 Threshold       | 13.4479| 18.6967 | 13.2169| 18.6953 | 22.0142| 22.0142 |
| Manhattan Precision          | 0.6178 | 0.5333  | 0.6172 | 0.5291  | 0.5891 | 0.5891  |
| Manhattan Recall             | 0.8226 | 0.9725  | 0.8089 | 0.9799  | 0.8902 | 0.8902  |
| Max AP                       | 0.6983 | 0.6481  | 0.6921 | 0.6471  | 0.6949 | 0.6949  |
| Max F1                       | 0.7134 | 0.6889  | 0.7045 | 0.6872  | 0.7128 | 0.7128  |
| Max Precision                | 0.6393 | 0.5347  | 0.6172 | 0.5291  | 0.6234 | 0.6234  |
| Max Recall                   | 0.8226 | 0.9725  | 0.8268 | 0.9926  | 0.8902 | 0.8902  |
| Similarity Accuracy          | 0.6687 | 0.6302  | 0.6622 | 0.6259  | 0.6703 | 0.6703  |
| Similarity Accuracy Threshold| 0.8435 | 0.7969  | 0.8553 | 0.7997  | 0.6514 | 0.6514  |
| Similarity AP                | 0.6983 | 0.6481  | 0.6921 | 0.6471  | 0.6949 | 0.6949  |
| Similarity F1                | 0.7134 | 0.6868  | 0.7045 | 0.6869  | 0.7128 | 0.7128  |
| Similarity F1 Threshold      | 0.8386 | 0.6921  | 0.8385 | 0.6684  | 0.5884 | 0.5884  |
| Similarity Precision         | 0.6393 | 0.5347  | 0.6136 | 0.5251  | 0.6234 | 0.6234  |
| Similarity Recall            | 0.8068 | 0.9599  | 0.8268 | 0.9926  | 0.8321 | 0.8321  |
| Main Score                   | 0.6983 | 0.6302  | 0.6921 | 0.6259  | 0.6949 | 0.6703  |
</details>

Samoed commented:
Results for clustering and pair classification very different

ayush1298 commented:
> Results for clustering and pair classification very different
>
Clustering are still in 1% limit, but for pair classification there are differences of 2% or sometimes more

ayush1298 commented:
> For `OCNLI` you have for `M1-Old` 0.6983 and `M1-New` 0.6302. This is 6%
>
yes, that's what I am seeing. Very high deviations in results there. 
So, what should we do now?

Samoed commented:
I'll try to look what can cause these changes

ayush1298 commented:
> I'll try to look what can cause these changes
>
@Samoed Did you get time to see what is going wrong here?

Samoed commented:
Not yet

Samoed commented:
@YanshekWoo can you share your evaluation script for mteb?

YanshekWoo commented:
> @YanshekWoo can you share your evaluation script for mteb? Created issue https://huggingface.co/HIT-TMG/KaLM-embedding-multilingual-mini-instruct-v1.5/discussions/10

For the instructions utilized in this study, interested researchers can refer to the corresponding code implementation available [here](https://github.com/HITsz-TMG/KaLM-Embedding/blob/f65ff7cfc5d15dabe03e3287ceeecc0d31b6007e/evaluate/utils.py#L121). 

Should you require any further assistance or have additional inquiries regarding this implementation, please do not hesitate to contact us.

ayush1298 commented:
@Samoed, I have run evaluations again 
When I run the following code once gain:
```
mteb run -m HIT-TMG/KaLM-embedding-multilingual-mini-instruct-v1.5 -t GeoreviewClusteringP2P
```

I getting the below warning:
```
Loader not specified for model HIT-TMG/KaLM-embedding-multilingual-mini-instruct-v1, loading using sentence transformers.
```
Though the loader is specified in its metadata.  Is this expected behaviour?
Then, below is next warning which says model prompts are not in expected format. But, we have added `prompts_dict=HIT_TMG_task_prompts`, so is it that we have to use `model_prompts` instead of `prompts_dict` here?
```
WARNING:mteb.models.sentence_transformer_wrapper:Model prompts are not in the expected format. Ignoring them.
```

Also, I got below warning:
```
INFO:mteb.models.sentence_transformer_wrapper:No model prompts found for task=GeoreviewClusteringP2P prompt_type=None
```
So, I think its using ` SentenceTransformerWrapper` instead of `InstructSentenceTransformerWrapper`, that may be causing all this issues, as all logging errors are from ` SentenceTransformerWrapper` .

Below is full trace when running above command: 
<details>

```

INFO:mteb.cli:Running with parameters: Namespace(model='HIT-TMG/KaLM-embedding-multilingual-mini-instruct-v1.5', task_types=None, categories=None, tasks=['GeoreviewClusteringP2P'], languages=None, benchmarks=None, device=None, output_folder='results', verbosity=2, co2_tracker=False, eval_splits=None, model_revision=None, batch_size=None, overwrite=False, save_predictions=False, func=<function run at 0x13d1c1000>)
INFO:mteb.models.overview:Model not found in model registry, assuming it is on HF Hub model.
INFO:mteb.models.overview:Attempting to extract metadata by loading the model (HIT-TMG/KaLM-embedding-multilingual-mini-instruct-v1.5) using HuggingFace.
WARNING:mteb.model_meta:Loader not specified for model HIT-TMG/KaLM-embedding-multilingual-mini-instruct-v1.5, loading using sentence transformers.
WARNING:mteb.models.sentence_transformer_wrapper:Model prompts are not in the expected format. Ignoring them.
INFO:mteb.evaluation.MTEB:

## Evaluating 1 tasks:
──────────────────────────────────────────────── Selected tasks  ─────────────────────────────────────────────────
Clustering
    - GeoreviewClusteringP2P, p2p


INFO:mteb.evaluation.MTEB:

********************** Evaluating GeoreviewClusteringP2P **********************
No config specified, defaulting to the single config: georeview-clustering-p2p/default
INFO:datasets.builder:No config specified, defaulting to the single config: georeview-clustering-p2p/default
Loading Dataset Infos from /Users/ayushmunot/miniconda3/envs/AyushOS/lib/python3.10/site-packages/datasets/packaged_modules/json
INFO:datasets.info:Loading Dataset Infos from /Users/ayushmunot/miniconda3/envs/AyushOS/lib/python3.10/site-packages/datasets/packaged_modules/json
Overwrite dataset info from restored data version if exists.
INFO:datasets.builder:Overwrite dataset info from restored data version if exists.
Loading Dataset info from /Users/ayushmunot/.cache/huggingface/datasets/ai-forever___georeview-clustering-p2p/default/0.0.0/97a313c8fc85b47f13f33e7e9a95c1ad888c7fec
INFO:datasets.info:Loading Dataset info from /Users/ayushmunot/.cache/huggingface/datasets/ai-forever___georeview-clustering-p2p/default/0.0.0/97a313c8fc85b47f13f33e7e9a95c1ad888c7fec
Found cached dataset georeview-clustering-p2p (/Users/ayushmunot/.cache/huggingface/datasets/ai-forever___georeview-clustering-p2p/default/0.0.0/97a313c8fc85b47f13f33e7e9a95c1ad888c7fec)
INFO:datasets.builder:Found cached dataset georeview-clustering-p2p (/Users/ayushmunot/.cache/huggingface/datasets/ai-forever___georeview-clustering-p2p/default/0.0.0/97a313c8fc85b47f13f33e7e9a95c1ad888c7fec)
Loading Dataset info from /Users/ayushmunot/.cache/huggingface/datasets/ai-forever___georeview-clustering-p2p/default/0.0.0/97a313c8fc85b47f13f33e7e9a95c1ad888c7fec
INFO:datasets.info:Loading Dataset info from /Users/ayushmunot/.cache/huggingface/datasets/ai-forever___georeview-clustering-p2p/default/0.0.0/97a313c8fc85b47f13f33e7e9a95c1ad888c7fec
INFO:mteb.abstasks.AbsTask:
Task: GeoreviewClusteringP2P, split: test, subset: default. Running...
INFO:mteb.models.sentence_transformer_wrapper:No model prompts found for task=GeoreviewClusteringP2P prompt_type=None
INFO:mteb.models.sentence_transformer_wrapper:Encoding 2000 sentences.
INFO:mteb.evaluation.MTEB:Evaluation for GeoreviewClusteringP2P on test took 159.70 seconds
INFO:mteb.evaluation.MTEB:Scores: {'default': {'v_measures': defaultdict(<class 'list'>, {'Level 0': [np.float64(0.6315550310927713), np.float64(0.6322401997311622), np.float64(0.6542310079717197), np.float64(0.6272222077187308), np.float64(0.6176991702060789), np.float64(0.6245929439556309), np.float64(0.638714267139654), np.float64(0.6315544690914952), np.float64(0.6343336194377835), np.float64(0.6314142212721889)]}), 'v_measure': 0.6323557137617215, 'v_measure_std': np.float64(0.009091107712532312), 'main_score': 0.6323557137617215}}

```

</details>

Also, results are increased from 61% to 63% for both `HIT-TMG/KaLM-embedding-multilingual-mini-instruct-v1` and `HIT-TMG/KaLM-embedding-multilingual-mini-instruct-v1.5` on running on `GeoreviewClusteringP2P`. 

For Ocnil, above 3 logging error are same. Except there is 1 more thing that may be causing large difference in results.
Ocnil doesn't have any prompt defined. Also, it has many duplicates, below is logger warning:
```
WARNING:mteb.evaluation.evaluators.PairClassificationEvaluator:Found 326/3694 duplicates in the input data. Only encoding unique sentences.
```
So, I am not sure here, whether the original evaluation done excludes these duplicates or not. 
Btw, on `Ocnil`, results of both `HIT-TMG/KaLM-embedding-multilingual-mini-instruct-v1` and `HIT-TMG/KaLM-embedding-multilingual-mini-instruct-v1.5` increase upto 66%. 

Can you answer these questions, I think this may be reason why we are not able to reproduce the results.

Samoed commented:
> Though the loader is specified in its metadata. Is this expected behaviour?

No, your loader haven't applied. I'll look why

Samoed commented:
I've run models and got these results. Prompts are applied 
```bash
mteb run -m HIT-TMG/KaLM-embedding-multilingual-mini-instruct-v1 -t Ocnli
```

```
Task: Ocnli, split: validation, subset: default. Running...
WARNING:mteb.evaluation.evaluators.PairClassificationEvaluator:Found 326/3694 duplicates in the input data. Only encoding unique sentences.
INFO:mteb.models.instruct_wrapper:Using instruction: 'Instruct: Retrieve semantically similar text.
Query: ' for task: 'Ocnli'
INFO:mteb.evaluation.evaluators.PairClassificationEvaluator:Computing similarity distances.
INFO:mteb.evaluation.evaluators.PairClassificationEvaluator:Computing metrics...
INFO:mteb.evaluation.MTEB:Evaluation for Ocnli on validation took 20.52 seconds
INFO:mteb.evaluation.MTEB:Scores: {'default': {'similarity_accuracy': 0.6302111532214402, 'similarity_accuracy_threshold': 0.7968777418136597, 'similarity_f1': 0.6868152625613902, 'similarity_f1_threshold': 0.6920686960220337, 'similarity_precision': 0.5347058823529411, 'similarity_recall': 0.9598732840549102, 'similarity_ap': 0.6481495013930183, 'cosine_accuracy': 0.6302111532214402, 'cosine_accuracy_threshold': 0.7968777418136597, 'cosine_f1': 0.6868152625613902, 'cosine_f1_threshold': 0.6920686960220337, 'cosine_precision': 0.5347058823529411, 'cosine_recall': 0.9598732840549102, 'cosine_ap': 0.6481495013930183, 'manhattan_accuracy': 0.625879805089334, 'manhattan_accuracy_threshold': 14.889359474182129, 'manhattan_f1': 0.6888556469708302, 'manhattan_f1_threshold': 18.69671630859375, 'manhattan_precision': 0.5332947307469601, 'manhattan_recall': 0.9725448785638859, 'manhattan_ap': 0.6458268546580729, 'euclidean_accuracy': 0.6302111532214402, 'euclidean_accuracy_threshold': 0.6373730897903442, 'euclidean_f1': 0.6868152625613902, 'euclidean_f1_threshold': 0.7847688794136047, 'euclidean_precision': 0.5347058823529411, 'euclidean_recall': 0.9598732840549102, 'euclidean_ap': 0.6481399210544279, 'dot_accuracy': 0.6302111532214402, 'dot_accuracy_threshold': 0.7968778014183044, 'dot_f1': 0.6868152625613902, 'dot_f1_threshold': 0.6920686960220337, 'dot_precision': 0.5347058823529411, 'dot_recall': 0.9598732840549102, 'dot_ap': 0.6481399210544279, 'max_accuracy': 0.6302111532214402, 'max_f1': 0.6888556469708302, 'max_precision': 0.5347058823529411, 'max_recall': 0.9725448785638859, 'max_ap': 0.6481495013930183, 'main_score': 0.6302111532214402}}
```

```bash
mteb run -m HIT-TMG/KaLM-embedding-multilingual-mini-instruct-v1.5 -t Ocnli
```
```
Task: Ocnli, split: validation, subset: default. Running...
WARNING:mteb.evaluation.evaluators.PairClassificationEvaluator:Found 326/3694 duplicates in the input data. Only encoding unique sentences.
INFO:mteb.models.instruct_wrapper:Using instruction: 'Instruct: Retrieve semantically similar text.
Query: ' for task: 'Ocnli'
INFO:mteb.evaluation.evaluators.PairClassificationEvaluator:Computing similarity distances.
INFO:mteb.evaluation.evaluators.PairClassificationEvaluator:Computing metrics...
INFO:mteb.evaluation.MTEB:Evaluation for Ocnli on validation took 20.57 seconds
INFO:mteb.evaluation.MTEB:Scores: {'default': {'similarity_accuracy': 0.625879805089334, 'similarity_accuracy_threshold': 0.7996639609336853, 'similarity_f1': 0.6868834490317867, 'similarity_f1_threshold': 0.6683769822120667, 'similarity_precision': 0.5251396648044693, 'similarity_recall': 0.9926082365364308, 'similarity_ap': 0.6470850460755918, 'cosine_accuracy': 0.625879805089334, 'cosine_accuracy_threshold': 0.7996639609336853, 'cosine_f1': 0.6868834490317867, 'cosine_f1_threshold': 0.6683769822120667, 'cosine_precision': 0.5251396648044693, 'cosine_recall': 0.9926082365364308, 'cosine_ap': 0.6470850460755918, 'manhattan_accuracy': 0.6215484569572279, 'manhattan_accuracy_threshold': 14.875192642211914, 'manhattan_f1': 0.6871529063309885, 'manhattan_f1_threshold': 18.69525718688965, 'manhattan_precision': 0.5290763968072976, 'manhattan_recall': 0.9799366420274551, 'manhattan_ap': 0.6444744323535153, 'euclidean_accuracy': 0.625879805089334, 'euclidean_accuracy_threshold': 0.6329866051673889, 'euclidean_f1': 0.6868834490317867, 'euclidean_f1_threshold': 0.8143991827964783, 'euclidean_precision': 0.5251396648044693, 'euclidean_recall': 0.9926082365364308, 'euclidean_ap': 0.6470850460755918, 'dot_accuracy': 0.625879805089334, 'dot_accuracy_threshold': 0.7996639013290405, 'dot_f1': 0.6868834490317867, 'dot_f1_threshold': 0.6683770418167114, 'dot_precision': 0.5251396648044693, 'dot_recall': 0.9926082365364308, 'dot_ap': 0.6470850460755918, 'max_accuracy': 0.625879805089334, 'max_f1': 0.6871529063309885, 'max_precision': 0.5290763968072976, 'max_recall': 0.9926082365364308, 'max_ap': 0.6470850460755918, 'main_score': 0.625879805089334}}

```

Samoed commented:
| | HIT-TMG/KaLM-embedding-multilingual-mini-instruct-v1 (PR) | HIT-TMG/KaLM-embedding-multilingual-mini-instruct-v1 (reported) |
|-|-|-|
| AskUbuntuDupQuestions | 0.62698 | 60.159 |
| SciDocsRR | | 80.722 |
| SCIDOCS | | 20.429 |
|SciFact | | 72.604 |

ayush1298 commented:
> I've run models and got these results. Prompts are applied. But it seems that prompts are not applied for `PairCLassifcation` tasks, at least in https://github.com/HITsz-TMG/KaLM-Embedding/blob/f65ff7cfc5d15dabe03e3287ceeecc0d31b6007e/evaluate/utils.py.
>
Yes, for `PairClassification`, no prompts are defined.
It seems that the loader is also getting correctly applied. Can you tell me what the issue was?
As I mentioned previously also, was having dupliacates was reason behind discrepencies in results, because in https://github.com/HITsz-TMG/KaLM-Embedding/blob/f65ff7cfc5d15dabe03e3287ceeecc0d31b6007e/evaluate/utils.py, I have not found anything that is handling duplicates, so maybe they have evaluated without dropping duplicates.

ayush1298 commented:
>
It seems that the loader is also getting correctly applied. Can you tell me what the issue was?
As I mentioned previously also, was having dupliacates was reason behind discrepencies in results, because in https://github.com/HITsz-TMG/KaLM-Embedding/blob/f65ff7cfc5d15dabe03e3287ceeecc0d31b6007e/evaluate/utils.py, I have not found anything that is handling duplicates, so maybe they have evaluated without dropping duplicates.
>
@Samoed Can you answer these, so I can wrap up this PR.

Samoed commented:
> Can you tell me what the issue was?

I don't know. This is hard to tell without information how you runned this model. Maybe you used different branch or something else. 

> I have not found anything that is handling duplicates

Duplicates are handled by pair classification evaluator

ayush1298 commented:
> I don't know. This is hard to tell without information how you runned this model. Maybe you used different branch or something else.
>
I have run models from this branch only. Also, how is it possible that with the same implementation, it is using the loader correctly for you but not for me?  

>Duplicates are handled by pair classification evaluator
>
So, what could be reason of difference in results then?

ayush1298 commented:
@Samoed, if everything works fine, when did you test, are you able to get results similar to the reported ones for `pairclassification`? The results in this [comment](https://github.com/embeddings-benchmark/mteb/pull/2478#issuecomment-2860186639) are before applying the instruction format correctly.

Samoed commented:
I've reevaluated and got `0.6323` on `Ocnli` for `HIT-TMG/KaLM-embedding-multilingual-mini-instruct-v1`

Samoed commented:
I've reevaluated and got `0.6323` on `Ocnli` for `HIT-TMG/KaLM-embedding-multilingual-mini-instruct-v1`

Samoed commented:
Hi @YanshekWoo! Can you help to find what can causes difference in results reevaluation of your models?

YanshekWoo commented:
> Hi @YanshekWoo! Can you help to find what can causes difference in results reevaluation of your models?
@Samoed 
I would be delighted to assist. 

Could you first help clarify the current issues requiring resolution?
1. Which specific benchmark(s) are you currently attempting to reproduce - MTEB and CMETB?
2. What particular datasets exhibit result differences?
3. Does the instruction mainly attribute to differences?

YanshekWoo commented:
For Ocnli, I reproduced the results of M2: HIT-TMG/KaLM-embedding-multilingual-mini-instruct-v1.5 under two sets of code.

| Parameters | Ocnli |
|  ----  | ----  |
| mteb==1.38.9, A100 | 0.662155 |
| mteb==1.38.9, T4 | 0.662155 |
| mteb==1.14.5, Ascend 910B (same as the [reproduction code](https://github.com/HITsz-TMG/KaLM-Embedding/tree/main) of reported results) | 0.679522038326835 |

Is it possible that the differences in results are due to different mteb versions (such as in score calculation or data)?
@ayush1298 @Samoed

Samoed commented:
I don't see much difference in code. I'll try to replicate it later, but your reported result is `69.214` and we can't reproduce it

YanshekWoo commented:
> I will conduct further tests in different machine environments to see if there are any differences in results under the same code.

We have also evaluate it on A100 with our original code. For `Ocnli`,  `M2: HIT-TMG/KaLM-embedding-multilingual-mini-instruct-v1.5` can also achieve `69.22`.  (https://github.com/embeddings-benchmark/mteb/pull/2478#issuecomment-2894166277 )

The core difference may stem from the MTEB version. 
Can I directly use the latest release of MTEB to reproduce your results?

YanshekWoo commented:
I have noticed that you set `apply_instruction_to_passages=True` in https://github.com/embeddings-benchmark/mteb/pull/2478/commits/e56ca1841f70b541722d74f6ed8f9a4a2691a620#diff-b6b40c2aeff83db93637d1270aa50058b0a901a3e6b95c5074d0acf03cb19979R169 , but our models do not add instruction for passages.
Could this potentially have an impact here?

Samoed commented:
> Could this potentially have an impact here?

Yes, this can be the reason. I'll reevaluate it. 

> Can I directly use the latest release of MTEB to reproduce your results?

No, you need to install mteb from this branch 

> Is there a complete script available that I can use directly?

```bash
mteb run -m HIT-TMG/KaLM-embedding-multilingual-mini-instruct-v1 -t Ocnli
```
Or
```python
import mteb 

model = mteb.get_model("HIT-TMG/KaLM-embedding-multilingual-mini-instruct-v1")
tasks = mteb.get_tasks(tasks=["Ocnli"])
evaluator = mteb.MTEB(tasks)
evaluator.run(model)
```

Samoed commented:
Hm. I tried to run it and got `0.62425`

YanshekWoo commented:
> mteb run -m HIT-TMG/KaLM-embedding-multilingual-mini-instruct-v1 -t Ocnli

Well. I will evaluate it with these scripts.

Samoed commented:
I've remembered that you haven't added prompts to `PairClassification` tasks, but mteb is adding it's default instruction for tasks. This describes why in this branch for pairclassfication we got `0.63`, but in main `0.66`.

I tried to run your script and got these results. They're maching with your reported scores, except for main score. It seems we had a bug for pair classification tasks, because there is not `max_accuracy`, which is `main_metric` and because of that you've reported `max_ap` as main score instead of `max_accuracy`
```
{'similarity_accuracy': 0.6621548456957228, 'similarity_accuracy_threshold': 0.8553438186645508, 'similarity_f1': 0.7044534412955467, 'similarity_f1_threshold': 0.8385477662086487, 'similarity_precision': 0.6136363636363636, 'similarity_recall': 0.8268215417106652, 'similarity_ap': 0.6921438509764521, 'cosine_accuracy': 0.6621548456957228, 'cosine_accuracy_threshold': 0.8553438186645508, 'cosine_f1': 0.7044534412955467, 'cosine_f1_threshold': 0.8385477662086487, 'cosine_precision': 0.6136363636363636, 'cosine_recall': 0.8268215417106652, 'cosine_ap': 0.6921438509764521, 'manhattan_accuracy': 0.6589063345966432, 'manhattan_accuracy_threshold': 12.571066856384277, 'manhattan_f1': 0.70018281535649, 'manhattan_f1_threshold': 13.216854095458984, 'manhattan_precision': 0.6172441579371475, 'manhattan_recall': 0.808870116156283, 'manhattan_ap': 0.689316563080095, 'euclidean_accuracy': 0.6621548456957228, 'euclidean_accuracy_threshold': 0.5378776788711548, 'euclidean_f1': 0.7044534412955467, 'euclidean_f1_threshold': 0.5682468414306641, 'euclidean_precision': 0.6136363636363636, 'euclidean_recall': 0.8268215417106652, 'euclidean_ap': 0.6921438509764521, 'dot_accuracy': 0.6621548456957228, 'dot_accuracy_threshold': 0.8553438186645508, 'dot_f1': 0.7044534412955467, 'dot_f1_threshold': 0.8385478258132935, 'dot_precision': 0.6136363636363636, 'dot_recall': 0.8268215417106652, 'dot_ap': 0.6921438509764521, 'max_f1': 0.7044534412955467, 'max_precision': 0.6172441579371475, 'max_recall': 0.8268215417106652, 'max_ap': 0.6921438509764521}
```

Samoed commented:
>  Is there any proper way to setting none instruction for STS and PairClassification under InstructSentenceTransformerWrapper class?

I think, yes we can change that. 


@ayush1298 can you evaluate these models on retrieval and reranking tasks to see if you need to apply prompt for passages, because it seems that we found main differences for tasks for now.

YanshekWoo commented:
> I tried to run your script and got these results. They're maching with your reported scores, except for main score. It seems we had a bug for pair classification tasks, because there is not `max_accuracy`, which is `main_metric` and because of that you've reported `max_ap` as main score instead of `max_accuracy`

I see. This may be due to a bug in the statistical methods used in previous versions of MTEB.

I have the JSON files of all previous evaluation results (mteb==1.14.5), and I can provide them if needed. Alternatively, they should also be directly available in the model's README.

Samoed commented:
No, I don't think we need to run tasks for now. We need to fix some problems with prompt
- Don't apply prompts on passages
- Don't apply prompts on PairClassification, STS and Summarization

Samoed commented:
> How can I ensure this thing?

You can see in logs which prompt is applied

Samoed commented:
@YanshekWoo I've updated implementation of your model and got these results for `HIT-TMG/KaLM-embedding-multilingual-mini-instruct-v1.5`:
| task | PR | Reported |
|-|-|-|
|SyntecRetrieval|0.81497|81.745|
|Ocnli|0.6621548456957228|69.2143|

YanshekWoo commented:
@Samoed  Got it. Thank you for your efforts.

Samoed commented:
I've run on more tasks `HIT-TMG/KaLM-embedding-multilingual-mini-instruct-v1.5` and got these results:
| task_name                          |   main_score | reported      |
|:-----------------------------------|-------------:| -------------:|
| EmotionClassification              |     0.86885  | 86.9 |
| SciFact                            |     0.71359  | 72.887|
| SciDocsRR                          |     0.816441 | 80.99314|
| SCIDOCS                            |     0.17654  | 19.967|
| STS16                              |     0.848328 | 83.5753|
| SprintDuplicateQuestions           |     0.930568 | 93.0568|
| ToxicConversationsClassification   |     0.892725 |89.277|
| AskUbuntuDupQuestions              |     0.60441  |60.3467|
| SummEval                           |     0.252182 |25.2273|
| TwitterSemEval2015                 |     0.714391 |71.4391|

There is most difference in retrieval tasks

Samoed commented:
Great! I think then we will wait for @KennethEnevoldsen review and after that we can merge it. Thank you for helping!

KennethEnevoldsen commented on mteb/models/hit_tmg_models.py:
```diff
@@ -0,0 +1,315 @@
+        instruction = self.get_task_instruction(
+            task_name, prompt_type, self.prompts_dict
+        )
+        from mteb import get_task
```
Is it here due to circular import?

Samoed replied:
I think yes

KennethEnevoldsen replied:
Alright - should we add a comment on this (ideally we should move it out in v2)

Samoed replied:
This problem won't occur in `v2`, because we're passing `TaskMetadata` directly

KennethEnevoldsen replied:
Perfect - I think we are good to merge then

KennethEnevoldsen reviewed (approved):
Generally things looks good here. I don't think I have a lot to add
